### PR TITLE
[iOS] Dispose of log readers and port forwarders if launch fails

### LINKED ISSUE #127141: Stuck waiting for Dart VM url with no stack trace
This issue is specifically for tracking test flakes when the test is stuck waiting for the Dart VM url and there is no stack trace or errors.

```
 [2023-05-16 20:19:59.651391] [STDOUT] stdout: [        ] (lldb)     connect
 [2023-05-16 20:19:59.673451] [STDOUT] stdout: [  +21 ms] (lldb)     run
 [2023-05-16 20:19:59.673515] [STDOUT] stdout: [        ] Debugger state set to launching.
 [2023-05-16 20:19:59.790472] [STDOUT] stdout: [ +116 ms] success
 [2023-05-16 20:19:59.790529] [STDOUT] stdout: [        ] Debugger state set to attached.
 [2023-05-16 20:19:59.790594] [STDOUT] stdout: [        ] Application launched on the device. Waiting for Dart VM Service url.
 ```
 
 Example: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20ios_platform_view_tests/10942/overview

### PR DESCRIPTION
**Pull request by vashworth**
When tests run in our CI using `flutter drive`, if there is a failure it will loop and try again.

https://github.com/flutter/flutter/blob/434b81f1a58b8528299a63ae7c3ded2aa519e3c9/packages/flutter_tools/lib/src/drive/drive_service.dart#L177-L186

However, it's using the same `device` instance for each iteration. So what was happening was when it would fail to launch, it would tell its listeners that it was cancelled.
https://github.com/flutter/flutter/blob/434b81f1a58b8528299a63ae7c3ded2aa519e3c9/packages/flutter_tools/lib/src/ios/ios_deploy.dart#L486-L489 

Then when the next iteration started, the `vmServiceDiscovery` would immediately return with null because the `deviceLogReader` would be cached from the previous iteration and would already be cancelled. Therefore, bypassing and cancelling the timer.

https://github.com/flutter/flutter/blob/434b81f1a58b8528299a63ae7c3ded2aa519e3c9/packages/flutter_tools/lib/src/ios/devices.dart#L585-L591

https://github.com/flutter/flutter/blob/434b81f1a58b8528299a63ae7c3ded2aa519e3c9/packages/flutter_tools/lib/src/ios/devices.dart#L627

In addition, it seems like sometimes the stop would fail and therefore the the drain would never get the signal that it was done and therefore would hang forever. There was no indication that the stop had failed though because the logs were going to the stream that had no listeners since `deviceLogReader` was already cancelled.
https://github.com/flutter/flutter/blob/434b81f1a58b8528299a63ae7c3ded2aa519e3c9/packages/flutter_tools/lib/src/ios/ios_deploy.dart#L563-L576

Fixes https://github.com/flutter/flutter/issues/127141

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 1]
```diff
@@ -587,12 +589,14 @@ class IOSDevice extends Device {
...
+        await dispose();
         return LaunchResult.failed();
       }
       return LaunchResult.succeeded(vmServiceUri: localUri);
     } on ProcessException catch (e) {
       await iosDeployDebugger?.stopAndDumpBacktrace();
       _logger.printError(e.message);
+      await dispose();
       return LaunchResult.failed();
     } finally {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/ios/devices.dart:601]: any reason not to just call `dispose()` from the `finally` block?

File: packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart [hunk 2]
```diff
@@ -147,6 +147,33 @@ void main () {
         expect(logger.traceText, contains('* thread #1'));
       });
 
+      testWithoutContext('debugger attached and stop failed', () async {
+        final StreamController<List<int>> stdin = StreamController<List<int>>();
+        final FakeProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+          FakeCommand(
+            command: const <String>['ios-deploy'],
+            stdout: '(lldb)     run\r\nsuccess\r\nsuccess\r\nprocess signal SIGSTOP\r\n\r\nerror: Failed to send signal 17: failed to send signal 17',
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:155]: Out of curiosity, would we actually expect to see `'\r'` here?

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 1 above)
Comment (vashworth) [reply packages/flutter_tools/lib/src/ios/devices.dart:601]: We don't want it to dispose when it launches successfully

File: packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart [hunk 3]
```diff
@@ -147,6 +147,33 @@ void main () {
         expect(logger.traceText, contains('* thread #1'));
       });
 
+      testWithoutContext('debugger attached and stop failed', () async {
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:150]: Is this test verifying any of your changes here? It's passing for me if I revert your other changes.

File: packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart (hunk 2 above)
Comment (vashworth) [reply packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:155]: Oh good question, I'm not sure. I just grabbed that from the test above. I'll try and figure out

File: packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart (hunk 3 above)
Comment (vashworth) [reply packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:150]: Oops, let me rework it

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart:601]: ahh yeah, good point :)

File: packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:155]: not worth investigating, just wondering if there was a legit reason.

File: packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart (hunk 3 above)
Comment (vashworth) [reply packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:151]: Okay updated it!

Comment (vashworth) [reply packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:151]: Okay updated it

**Review by christopherfujino — APPROVED** [review]
Nice! Thanks for this fix!

File: packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart (hunk 2 above)
Comment (vashworth) [reply packages/flutter_tools/test/general.shard/ios/ios_deploy_test.dart:156]: I just happened to come across the answer to this: https://github.com/flutter/flutter/blob/4f6f1887c5e0b610a906c1b3eb5f01a8d2a555bc/packages/flutter_tools/lib/src/ios/ios_deploy.dart#L477-L480

**Pull request merged**